<h3 style="color: #001a79;">Exercise 1.1</h3>

<hr style="border-top: 1px solid #001a79;" />


The above gives about a 1.5% chance of randomly selecting the correct cups. Calculate the minimum number of cups of tea required to ensure the probability of randomly selecting the correct cups is less than or equal to 1%.



Reference: https://www.w3schools.com/python/ref_math_comb.asp

If the number of ways of selecting 4 cups from 8, results in 70 combinations and has a chance of randomly selecting the correct 4 cups of approximately 1.5%, the next step will be trying to choose 4 cups from 9.

In [1]:
# Finding the amount of the combinations possible using the math module
import math

In [2]:
math.comb(9,4)

126

We can see that number of ways of selecting 4 cups from 9 will result in a total of <b>126</b> combinations. Now we can check the probability of selecting the correct 4 cups.

In [3]:
# Probability of random selecting 4 cups from 9:

1.0 / math.comb(9,4)

0.007936507936507936

As seen above, the probability of randomly selecting the 4 cups from a pool of 9 cups is approximately <b>0.7%<b>, which is less than 1%

<b>*Bonus:*</b> How many would be required if you were to let the taster get one cup wrong while maintaining the 1% threshold?

Since the number of combination needs to be higher than 100, to maintain the 1% threshold. if we just reduce the amount of right cups from 4 to 3 that needs to be correct it will result in a total of 84 combinations and a approximately 1.1% chance of the right cups being chosen at random.

In [4]:
math.comb(9,3)

84

In [5]:
1.0 / math.comb(9,3)

0.011904761904761904

So if a 3 cups from a 9 cups pool would result in a probability of random guessing the right cups higher than 1%, the next step would be increase the amount of cups that need to be selected. So instead of 9 we should try it with 10 cups.

In [6]:
math.comb(10,3)

120

In [7]:
1.0 / math.comb(10,3)

0.008333333333333333

As we can see above, if we increase the amount to 10 cups, it will result in a 120 combinations and a probability of random chosing the right cups of approximately 0.8%

<h3 style="color: #001a79;">Exercise 1.2</h3>


Use <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fisher_exact.html" style="color: #ff791e">scipy's version of Fisher's exact test</a> to simulate the Lady Tasting Tea problem.

In [8]:
import scipy.stats as ss

In [9]:
ss.fisher_exact([[4, 0], [0, 4]], alternative='greater')

(inf, 0.014285714285714268)

<h3 style="color: #001a79;">Exercise 1.3</h3>

Take the code from the <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html" style="color: #ff791e">Examples section of the scipy stats documentation for independent samples t-tests</a>, add it to your own notebook and add explain how it works using MarkDown cells and code comments. Improve it in any way you think it could be improved.

In [10]:
# Importing packges such as scipy and numpy and assingning the random generation function to the variable rng
from scipy import stats
import numpy as np
rng = np.random.default_rng()

When testing samples with identical means, the t and p values are the same.


In [11]:
# Generte random of array 500 numbers with the mean equal to 5 a standard variation of 10 
rvs1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
rvs2 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)

#Performs a t test on the arrays generated above
stats.ttest_ind(rvs1, rvs2)

Ttest_indResult(statistic=0.9468717378006166, pvalue=0.34393334228185646)

Setting the equal_var paramenter to false does not change the results since the actual variations is the same for both populations

In [12]:
# Setting equal_var as False does not assume equal variance between the samples
stats.ttest_ind(rvs1, rvs2, equal_var=False)

Ttest_indResult(statistic=0.9468717378006166, pvalue=0.34393357876557207)

ttest_ind underestimates p for unequal variances:

In [13]:
#Creating a third array of random numbers with the same mean as the others but larger variance of 20 and perfoming a t test 
rvs3 = stats.norm.rvs(loc=5, scale=20, size=500, random_state=rng)
stats.ttest_ind(rvs1, rvs3)

Ttest_indResult(statistic=-0.40460800218329795, pvalue=0.6858523750938635)

In [22]:
# Setting equal_var as False does not assume equal variance between the samples
stats.ttest_ind(rvs1, rvs3, equal_var=False)

Ttest_indResult(statistic=-0.40460800218329795, pvalue=0.685891102424002)

When the size of the populations is different the t and p value also change.

In [15]:
rvs4 = stats.norm.rvs(loc=5, scale=20, size=100, random_state=rng)
stats.ttest_ind(rvs1, rvs4)

Ttest_indResult(statistic=2.7027201870593935, pvalue=0.007072827481774469)

In [16]:
stats.ttest_ind(rvs1, rvs4, equal_var=False)

Ttest_indResult(statistic=1.6594488046669282, pvalue=0.09996074602930462)

Perfoming the test in populations with different means, variances and sizes.

In [17]:
rvs5 = stats.norm.rvs(loc=8, scale=20, size=100, random_state=rng)
stats.ttest_ind(rvs1, rvs5)

Ttest_indResult(statistic=1.413436394006248, pvalue=0.15804776927780065)

In [18]:
stats.ttest_ind(rvs1, rvs5, equal_var=False)

Ttest_indResult(statistic=0.8302918372110254, pvalue=0.40825239491973675)

In this case P value is more accurate in this case because the test was peformed several times, using the numpy random function that was assingned to the variable rng.

In [19]:
# Added a permutation parameter to produce the test x amount of times.
stats.ttest_ind(rvs1, rvs5, permutations=10000,
                random_state=rng)

Ttest_indResult(statistic=1.413436394006248, pvalue=0.1602)

Take these two samples, one of which has an extreme tail.

In [20]:
a = (56, 128.6, 12, 123.8, 64.34, 78, 763.3)
b = (1.1, 2.9, 4.2)

Using the trim parameter to "cut" elements from the population, in this case it was used .2. The calculation of how many elements will be trimmed is the floor value of the multiplication between the length of the array and the trim parameter. In this case for "a" it will be np.floor(7*0.2) which is equal to 1. And for "b" it will be np.floor(3*0.2) which is equal to zero. 

In [24]:
# trimmed values for a
trim_a = np.floor(len(a)*.2)
trim_a

1.0

In [25]:
# trimmed values for b
trim_b = np.floor(len(b)*.2)
trim_b

0.0

In [21]:
stats.ttest_ind(a, b, trim=.2)

Ttest_indResult(statistic=3.4463884028073513, pvalue=0.01369338726499547)